---
title: "VGAM single donor"
output: html_notebook
---

```{r}
source("src/ariana_round2_VGAM.R")
```

```{r}
vcf_file <- "/home/dmeyer/projects/bqtls/SecondRound_bQTLs/VCF_files/bgzipped_vcfs/microglia_SingleDonor.vcf.gz"
feather <- "/home/dmeyer/projects/bqtls/SecondRound_bQTLs/VCF_files/feathers/microglia_SingleDonor.feather"
obs_files <- list.files("/home/dmeyer/projects/bqtls/SecondRound_bQTLs/Microglia_Single_Donor_PU1/", "*allelic_out.txt$", full.names = TRUE)
tfs <- "PU1"
out_files <- paste0("~/projects/bqtls/SecondRound_bQTLs/asb/Microglia_Single_Donor_", tfs, ".model_results.txt")
deconv_out_files <- paste0("/home/dmeyer/projects/bqtls/SecondRound_bQTLs/deconvolve_data/Microglia_Single_Donor_", tfs, ".deconvolution.txt")
cell_lines <- get_cell_lines_from_vcf(vcf_file)
geno <- read_feather_file(feather, cell_lines)%>%mutate(AC = AC*2)
pdfs <- str_replace(out_files, "model_results.txt$", "model_output.pdf")
    
#geno defined
# cell_lines defined
obs_file <- obs_files[1]
out_file <- out_files[1]

obs <- fread(obs_file, sep = "\t", header = F, skip = 1) 
obs <- obs[, 1:(ncol(obs) - 1)]
colnames(obs) <- unlist(strsplit(readLines(obs_file, 1), "\t"))

if (!all(obs$variantID %in% geno$variantID)) {
    stop("Allele flipping in", obs_file)
}
#if (!all(obs$variantID %in% c(geno$varinatID, geno_flipped$variantID))) {
#    stop("Not all variants observed in CHIP-seq in VCF")
#}
```
```{r}
geno$AC <- geno$AC * 2
```

Question: Is there any reason that this pipeline would do poorly for
`altAllele = TTATTTT` (indels) compared to single-nucleotide variants?
```{r}
head(obs)
head(geno)
```

```{r}
hist(geno$AC)
```


```{r}
total_thresh <- 30
obs_sub <- obs[(totalCount >= total_thresh) & (totalCount != refCount) & (totalCount != altCount)]
# MAF threshold doesn't make sense here but we definitely only want HET SNPs
geno_sub <- geno[AC == 1]
variants_keep <- intersect(obs_sub$variantID, geno_sub$variantID)
obs_sub <- obs_sub[match(variants_keep, variantID)]
geno_sub <- geno[match(variants_keep, variantID)]
X <- as.matrix(geno[match(variants_keep, variantID), ..cell_lines])

plot(table(geno$AC), main = paste(nrow(X), "variants retained with AC >= 0"))
abline(v=0.5, col='red', lty='dashed')
abline(v=1.5, col='red', lty='dashed')
```


```{r}
# observed allelic ratios
afs_expected <- geno_sub$AF
observed_afs <- with(obs_sub, altCount / totalCount)

p2 <- ggplot(data.frame(afs_expected, observed_afs), aes(x = afs_expected, y = observed_afs, color = factor(afs_expected))) + 
    geom_point(alpha = 0.4, position = position_jitterdodge(jitter.width = 0.4), show.legend = F) +
    geom_violin(alpha = 0.8, show.legend = F)+
    #geom_smooth()+
    labs(x = "Allelic ratio if no ASB", y = "Observed allelic ratio")+
    theme_linedraw()
p2
```

```{r}
hist(observed_afs, main = "Histogram of observed AFs in ChIP-seq data\nExpected = 0.5 (red line), Mean = "%>%paste0(round(mean(observed_afs),5)))
abline(v=0.5, col = 'red', lty='dashed', lwd=3)
```


```{r}
hist(log10(observed_afs), main="hist of log10(obs_af) mean = "%>%paste0(round(mean(log10(observed_afs)),3)))
```

```{r}
any(is.na(afs_expected))
any(is.na(obs_sub$altCount))
any(is.na(obs_sub$refCount))
hist(afs_expected)
afs_expected <- ifelse(afs_expected == 1, 1 - .Machine$double.eps, afs_expected)
```


```{r}
#fit <- vglm(cbind(altCount, refCount) ~ identitylink(offset(mu)), betabinomial(lmu = identitylink, lrho = identitylink), obs_sub, crit = "coef")
fit <- vglm(cbind(altCount, refCount) ~ 1, betabinomial(lmu = identitylink, lrho = identitylink), obs_sub, crit = "coef")
#fit <- vglm(cbind(altCount, refCount) ~ identitylink(offset(0.5)), betabinomial(lmu = identitylink, lrho = identitylink), obs_sub, crit = "coef")
```

```{r}
cat("Coefficients of fit")
print(coef(fit, matrix = TRUE))

rho <- coef(fit, matrix = TRUE)[1,2]
conc <- (1-rho)/rho
print(paste0("conc = ",conc))
```

```{r}
i = 3634
crit1 <- obs_sub$totalCount[i]*0.5
crit2 <- obs_sub$altCount[i]
if (is.na(crit1) | is.na(crit2)) {
    next()
}
if (crit2 > crit1) {
    pbetabinom(q = obs_sub$altCount[i], size = obs_sub$totalCount[i], prob = 0.5, rho = rho)
    pval =1-pbetabinom(q = obs_sub$altCount[i]-1, size = obs_sub$totalCount[i], prob = 0.5, rho = rho)
    .title = paste0("P(x > ", round(crit2,2), ") = ", format(pval, scientific = TRUE, digits = 3))
} else {
    pval = pbetabinom(q = obs_sub$altCount[i], size = obs_sub$totalCount[i], prob = 0.5, rho = rho)
    .title = paste0("P(x <= ", round(crit2,2), ") = ", format(pval, scientific = TRUE, digits = 3))
}

p=rbetabinom(9000, obs_sub$totalCount[i], 0.5)
hist(p, main = .title, xlim = c(min(c(crit2, min(p))), max(crit2, max(p))))
#plot(density(p, from = 0, width=3), main = .title, xlim = c(min(c(crit2, min(p))), max(crit2, max(p))))
abline(v = crit1, col = 'red', lty='dashed')
abline(v = crit2, col = 'blue', lty='dashed')
```

```{r}
qvals <- p.adjust(pvals, method = "fdr")
hist(-log10(qvals))
abline(v = -log10(0.05), col='red', lty='dashed')
data.frame(i = 1:length(qvals), q = qvals)%>%
  arrange(q)
```


```{r}
pvals <- sapply(seq(1,nrow(obs_sub)), function(i) {
    crit1 <- obs_sub$totalCount[i]*0.5
    crit2 <- obs_sub$altCount[i]
    if (is.na(crit1) | is.na(crit2)) {
       return(NA)
    }
    p1 = 1-pbetabinom(q = obs_sub$altCount[i]-1, size = obs_sub$totalCount[i], prob = 0.5, rho = rho)
    p2 = pbetabinom(q = obs_sub$altCount[i], size = obs_sub$totalCount[i], prob = 0.5, rho = rho)
    pval = 2* min(p1, p2)
    return (pval)
})
```

```{r}
hist(pvals)
```

```{r}
hist(-log10(pvals))
```

```{r}
results <- mutate(obs_sub,p_value = pvals,
    direction = sapply(1:nrow(obs_sub), function(i) {
        crit1 <- obs_sub$totalCount[i]*0.5
        crit2 <- obs_sub$altCount[i]
        if (is.na(crit1) | is.na(crit2)) {
           return(NA)
        }
        if (crit2 > crit1) {
            return("Positive")
        } else { 
            return("Negative")
        }
    }))%>%
arrange(p_value)
write_tsv(results, "~/projects/bqtls/SecondRound_bQTLs/asb/Microglia_Single_Donor_PU1.model_results.txt")
```



Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
